### DISCUSSION TITLE
functools.partial() does not work on an async dependable

### DISCUSSION THREAD

**Discussion by lieryan** [fastapi/fastapi#9744] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from functools import partial

from fastapi import FastAPI, Depends


app = FastAPI()

def the_answer():
    return 42


async def dep1(
    n: int = Depends(the_answer),
):
    return n + 1

@app.get("/")
def main(
    d1=Depends(partial(dep1, n=10)),
):
    print(d1)   # prints "<coroutine object dep1 at 0x7facc996fb50>"
    return {}


if __name__ == '__main__':
    import uvicorn
    uvicorn.run(app, port=1122)
```

### Description

I would've expected `d1` above to be the return value of the dependable, but it seems like fastapi didn't identify the partial as an coroutine function and instead passed the unawaited coroutine into the view function.

`inspect.iscoroutinefunction()` and `inspect.signature()` were able to deconstruct the partial object to correctly identify whether a partial contains a coroutine or a regular function, so it seems like it should've been possible for FastAPI to use handle partial correctly:

```
In [5]: def regular(a, b): pass

In [6]: async def coro(a, b): pass

In [7]: inspect.iscoroutinefunction(partial(regular, 1))
Out[7]: False

In [8]: inspect.iscoroutinefunction(partial(coro, 1))
Out[8]: True

In [20]: inspect.signature(partial(coro, 1))
Out[20]: <Signature (b)>
```
 

### Operating System

Linux

### FastAPI Version

0.98.0

### Python Version

Python 3.10.6

**Comment by Kludex** [comment]
Yeah. I think FastAPI would need to unwrap the `partial`, like it's being done for everything else in Starlette: https://github.com/encode/starlette/blob/master/starlette/_utils.py#L14-L15

**Reply by Kludex** [reply]
There was a PR that was closed some time ago: https://github.com/tiangolo/fastapi/pull/3343

**Reply by lieryan** [reply]
I've written an updated PR that would implement support for this and other uses of partials on the various forms of dependables: #9753. Unlike #3343, it also ensures that partial will also work reliably in all sorts of async/callable dependables, in addition to `async def`.

**Reply by realitycheck** [reply]
Any chances of #9753 being merged and released soon?

**Comment by Kludex** [comment, marked as answer]
I've created this: https://github.com/tiangolo/fastapi/issues/10720
